### ISSUE TITLE
File modification events are getting lost in PollingWatchService

### ISSUE THREAD

**Issue by mechkg** [sbt/io#86]
While investigating https://github.com/sbt/sbt/issues/3775 I think I have narrowed the bug down to the interaction between `PollingWatchService.pollEvents` and `PollingWatchService.PollingThread.addEvent`. 

Events often get lost without being properly consumed (https://github.com/sbt/io/issues/82 is probably related). 

**UPDATE**: the conclusion below is wrong, see comments further down.

As far as I understand the only reason that could happen is if some events get added between these two statements: 

https://github.com/sbt/io/blob/e8c9757799bb0b251e5c5aefd96f841428d0eb9a/io/src/main/scala/sbt/io/PollingWatchService.scala#L61-L64

so `thread.keysWithEvents.clear` deletes events that weren't consumed. This should be prevented by the `synchronized` guard.

It is difficult to set up my own modified sbt build to properly test this but at a glance I find this line suspicious:

https://github.com/sbt/io/blob/e8c9757799bb0b251e5c5aefd96f841428d0eb9a/io/src/main/scala/sbt/io/PollingWatchService.scala#L105

I don't understand the purpose of the `notifyAll` call at this point and I suspect that it could break the intended synchronization behaviour.

**Comment by dwijnand** [comment]
Thank you for the insight @mechkg, I'll look into what you've found (it works well with https://github.com/sbt/sbt/issues/3687, https://github.com/sbt/sbt/issues/3695, https://github.com/sbt/io/issues/82, and https://github.com/sbt/sbt/issues/3775 which I'm look at).

**Comment by mechkg** [comment]
In `PollingWatchService.pollEvents` events are gathered as follows:

```
   val events = thread.keysWithEvents.map { k =>
        k -> k.pollEvents().asScala.asInstanceOf[Seq[WatchEvent[JPath]]]
      }
```

`PollingWatchKey.pollEvents` in turn does this:

```
  origin.keysWithEvents -= this
```

So keysWithEvents is being modified by `PollingWatchKey.pollEvents` at the same time as it's being mapped over in `PollingWatchService.pollEvents` which looks very dangerous. Usually in Java this would result in a `ConcurrentModificationException` but it does not in this case, I wonder why?

**Comment by mechkg** [comment]
Simply removing this line:

     origin.keysWithEvents -= this

apparently makes the problem go away. It looks like it is not necessary because `keysWithEvents` is getting cleared at the end of `PollingWatchService.pollEvents` anyway, but I am not sure if any other code depends on this behaviour?

**Comment by dwijnand** [comment]
Fixed with #90.
